src/Features/Core/Portable/Diagnostics/EngineV2/CompilerDiagnosticExecutor.cs

```diff
@@ -0,0 +1,292 @@
+                    switch (diagnostic.Location.Kind)
+                    {
+                        case LocationKind.ExternalFile:
+                            {
+                                var diagnosticDocumentId = GetExternalDocumentId(diagnostic);
+                                if (documentId == diagnosticDocumentId)
+                                {
+                                    var document = _project.GetDocument(diagnosticDocumentId);
+                                    if (document != null)
+                                    {
+                                        // local diagnostics to a file
+                                        lazyLocals = lazyLocals ?? new Dictionary<DocumentId, List<DiagnosticData>>();
+                                        lazyLocals.GetOrAdd(document.Id, _ => new List<DiagnosticData>()).Add(DiagnosticData.Create(document, diagnostic));
+
+                                        SetDocument(document);
+                                    }
+                                }
+                                else if (diagnosticDocumentId != null)
+                                {
+                                    var document = _project.GetDocument(diagnosticDocumentId);
+                                    if (document != null)
+                                    {
+                                        // non local diagnostics to a file
+                                        _lazyNonLocals = _lazyNonLocals ?? new Dictionary<DocumentId, List<DiagnosticData>>();
+                                        _lazyNonLocals.GetOrAdd(document.Id, _ => new List<DiagnosticData>()).Add(DiagnosticData.Create(document, diagnostic));
+
+                                        SetDocument(document);
+                                    }
+                                }
+                                else
+                                {
+                                    // non local diagnostics without location
+                                    _lazyOthers = _lazyOthers ?? new List<DiagnosticData>();
+                                    _lazyOthers.Add(DiagnosticData.Create(_project, diagnostic));
+                                }
+
+                                break;
+                            }
+                        case LocationKind.None:
+                            {
+                                _lazyOthers = _lazyOthers ?? new List<DiagnosticData>();
+                                _lazyOthers.Add(DiagnosticData.Create(_project, diagnostic));
+                                break;
+                            }
+                        case LocationKind.SourceFile:
+                        case LocationKind.MetadataFile:
+                        case LocationKind.XmlFile:
+                            {
+                                // something we don't care
+                                continue;
```

Comment:
Why can't these have LocationKind.SourceFile? Even if the project they are analyzing don't support Compilation, can't they report diagnostics with location as SourceFile for documents in the project?